pkg/util/containers/v2/metrics/provider/collector_cache.go

```diff
@@ -80,56 +73,64 @@ func (cc *collectorCache) GetContainerNetworkStats(containerID string, cacheVali
 	currentTime := time.Now()
 	cacheKey := contNetStatsCachePrefix + containerID
 
-	if cacheValidity > 0 {
-		entry, found, err := cc.getCacheEntry(currentTime, cacheKey, cacheValidity)
-		if found {
-			if err != nil {
-				return nil, err
-			}
-
-			return entry.(*ContainerNetworkStats), nil
+	entry, found, err := cc.cache.Get(currentTime, cacheKey, cacheValidity)
+	if found {
+		if err != nil {
+			return nil, err
 		}
+
+		return entry.(*ContainerNetworkStats), nil
 	}
 
 	// No cache, cacheValidity is 0 or too old value
 	val, err := cc.collector.GetContainerNetworkStats(containerID, cacheValidity)
 	if err != nil {
-		cc.storeCacheEntry(currentTime, cacheKey, nil, err)
+		cc.cache.Store(currentTime, cacheKey, nil, err)
 		return nil, err
 	}
 
-	cc.storeCacheEntry(currentTime, cacheKey, val, nil)
+	cc.cache.Store(currentTime, cacheKey, val, nil)
 	return val, nil
 }
 
-func (cc *collectorCache) getCacheEntry(currentTime time.Time, key string, cacheValidity time.Duration) (interface{}, bool, error) {
-	cc.cacheLock.RLock()
-	entry, found := cc.cache[key]
-	cc.cacheLock.RUnlock()
+// GetContainerIDForPID returns the container ID for given PID
+// errors are cached as well to avoid hammering underlying collector
+func (cc *collectorCache) GetContainerIDForPID(pid int, cacheValidity time.Duration) (string, error) {
```

Comment:
So `collectorCache` provides the generic caching mechanism, I don't think we should change that.
There is indeed two layers of caching, the `metadataCollector` being cached on top of the collector, it was done on purpose to avoid repeating the process of going through this process too often:
```
for each collector -> collector.Get() -> Force a refresh -> not found
```

But I guess that in most cases we'll only have 1 collector, and we should probably optimize for the most common use case, so I can remove the `collectorCache` on top of `metaCollector`